### DISCUSSION TITLE
How to detect and transfer Mesh, Materials etc. across network

### DISCUSSION THREAD

**Discussion by raffaeleragni** [bevyengine/bevy#8815] [Q&A] [answered]
Hi again.

## Context

I am in the process of serializing data across network with my plugin. I have entities and components covered by now so I am moving on to asset types. This includes Mesh, Materials, and all other assets either created programmatically or imported from the asset folder.

I noticed that when I sync the Handle component between clients it automatically gets replicated with an already filled reference to the same asset id or asset path id. So in theory all I need to do is a plugin that keeps assets in sync directly from memory, and clients would just be able to reference the same id. This assumes that asset ids are deterministic, and/or I can create an asset with a specified id on clients.

Also StandardMaterial is Reflect and that is okay, but for example Mesh is not even Reflect.

## Question

What is the intended way from the engine authors to handle these types for these situations? And to be able essentially to create a custom asset manager?

Mesh & StandardMaterial are not components either, but assets used from Handle components. How to detect their change, eg if they get reloaded from disk, since they won't be part of the ECS essentially, I assume I cannot just query them? Is there another plugin I have to hook into in order to know when these are newly added or changed, or deleted?

**Comment by raffaeleragni** [comment]
Is by any chance `AssetEvent` relevant to this?

**Comment by raffaeleragni** [comment]
I am trying to advance on this currently. Right now I am at the stage of creating a test setting up bevy for a minimal + assets enabled.

However it seems that `MinimalPlugins` won't support assets (No Assets resource), and I can't use `DefaultPlugins` either because they will require a window/render and make automated tests not possible, as not able to control manual update() calls anymore.

How to setup a MinimalPlugins configuration that will also have Assets enabled?

**Comment by raffaeleragni** [comment]
Can anybody help here, in how to setup a MinimalPlugin with assets (mesh, materials) available?

**Comment by nicopap** [comment, marked as answer]
What do you mean by "sync asset ID between clients"

> I noticed that when I sync the Handle component between clients it automatically gets replicated with an already filled reference to the same asset id or asset path id. 

Do you mean several entities in the same `World` as in: within a single process? (eg: you lauched the game once by running the compiled the binary)

Are you asking how to extend this to multiple processes or multiple devices? This is not how assets are supposed to work. I don't understand what you mean.

### How to create your own asset handling?

You would implement your own [`AssetIo`](https://docs.rs/bevy/latest/bevy/asset/trait.AssetIo.html). For example, for WASM, bevy has a different, http-based, `AssetIo` implementation. With `AssetIo` you can use arbitrary logic to load assets.

### How to add assets to a `MinimalPlugin`?

You would manually add the `bevy::asset::AssetPlugin` to your `app`, in addition to the `MinimalPlugin`, similarly to any plugin.

As follow: `.add_plugin(bevy::asset::AssetPlugin::default())`

**Comment by raffaeleragni** [comment]
I'm writing a network plugin aimed at collaborative editing to keep things in sync across different clients. Actions in one client will be propagated across all the others. Since I've covered components for now, `Handle` is being transferred like this:

![image](https://github.com/bevyengine/bevy/assets/402368/e4d6ef82-6185-406b-85b4-a3ed73cdcfdc)

So now I have to address actual assets transfer through the network 

In order to develop things right, I am creating automated tests that will check this behavior and in doing so I am setting up bevy with minimal plugins (because headless, for automation).

While adding `AssetPlugin` on top of `MinimalPlugins` and then trying to setup, let's say, a `StandardMaterial` + `Mesh` assets, this happens: `Requested resource bevy_asset::assets::Assets<bevy_render::render_resource::shader::Shader> does not exist in the World`. I have tried to add more plugins so far following what `DefaultPlugins` is doing, but tests should be simpler than initializing these many things. (So far I have also added: `PbrPlugin`, `ScenePlugin`, `RenderPlugin`, `ImagePlugin`, `PipelinedRenderingPlugin`, `CorePipelinePlugin`. But without success).

While my original question is about Assets, and it still keeps being so, I am also starting to worry about how bevy will address these test environments situations, that is, test setups that are not exclusively minimal. I'm pretty sure at some point these will surface themselves too.

**Comment by raffaeleragni** [comment]
After some lenghty trial, I ended up with this setup to have a StandardMaterial in a test environment:
```rust
    app.add_plugins(MinimalPlugins);
    app.add_plugin(AssetPlugin::default());
    app.add_asset::<Shader>().add_debug_asset::<Shader>();
    app.add_plugin(PbrPlugin::default());
```

I think I can continue with my work for now, I will close this question and go with the AssetPlugin solution with this modification.
